### ISSUE TITLE
Anomaly: Evar under binder causes _UNBOUND_REL after applying lemma

### ISSUE THREAD

**Issue by fakusb** [coq/coq#13363]
#### Description of the problem

Applying a Lemma to a goal where an evar is under a binder causes an anomaly:

The following has `P (fun _ : X => ?ss'@{y:=_UNBOUND_REL_2})` as proof state before `cbn beta`, and finally fails at `cbn beta` with `Anomaly "Uncaught exception Not_found."`

````
Axiom X : Type.       
Axiom P : (X -> unit) -> Prop.

Axiom F: unit -> unit.
Axiom G : unit -> unit.

Lemma Hyp ss':
    P (fun y : X => ss') ->
    P (fun y : X => G (F ss')).
Admitted.

Lemma bug : exists x, P x.
Proof.
intros.
eexists (fun y : X => G _).
eapply Hyp. cbn beta.
Qed.
````

I would expect the `eapply` to behave like `refine (Hyp _ _).`
(In my real-world example, I would also like to keep `y` available in the evar in the hypothesis of the lemma, although this is not reflected in this minimal example)

#### Coq Version

8.12 and 8.13 build yesterday.

**Comment by fakusb** [comment]
For the uninformed (me): I think the apply is buggy, not the cbn. Is this nevertheless a reduction-strategies problem, @herbelin ?

**Comment by herbelin** [comment]
> For the uninformed (me): I think the apply is buggy, and the failing cbn only is a symptom. Is this nevertheless a reduction-strategies problem, @herbelin ?

Ah right, I was too quick. Seems instead like a restriction problem in unification. The evar argument of `G` (named `?Goal` at this time should not be allowed to depend on `y` any more when it becomes `?ss'`.

**Comment by herbelin** [comment]
This is then fixed by #13373. Thanks a lot for the report!
